File: frontend/src/api/quiz.ts
import { apiClient, ApiEnvelope } from "./client";

export type QuestionType = "mcq" | "numerical" | "short_answer";
export type QuizMode = "practice" | "timed_assessment";
export type AssessmentType = "topic_assessment" | "chapter_test" | "comprehensive_exam" | "diagnostic_test";
export type DifficultyLevel = "easy" | "medium" | "hard" | "adaptive";

export interface QuizOption {
  id: string;
  text: string;
}

export interface QuizQuestionPublic {
  id: string;
  question_type: QuestionType;
  prompt: string;
  options?: QuizOption[] | null;
  topic_title?: string;
  difficulty?: DifficultyLevel;
}

export interface QuizResponse {
  id: string;
  profile_id: string;
  roadmap_node_id?: string | null;
  mode: QuizMode;
  total_questions: number;
  time_limit_seconds?: number | null;
  difficulty?: string;
  questions: QuizQuestionPublic[];
}

export interface QuizUserAnswer {
  question_id: string;
  user_answer: string;
}

export interface QuestionResult {
  question_id: string;
  question_type: QuestionType;
  prompt: string;
  user_answer: string;
  correct_answer: string;
  is_correct: boolean;
  score: number;
  error_category?: string;
  feedback: string;
  explanation: string;
  remediation_advice?: string;
}

export interface MasteryNodeUpdate {
  node_id: string;
  node_title: string;
  previous_mastery: number;
  new_mastery: number;
  status: string;
  unlocked: boolean;
}

export interface QuizResultResponse {
  id: string;
  profile_id: string;
  roadmap_node_id?: string | null;
  /** Resolved server-side from the roadmap node. */
  topic_title?: string | null;
  mode: string;
  started_at: string;
  completed_at?: string | null;
  score: number;
  letter_grade?: string;
  total_questions: number;
  correct_count: number;
  time_limit_seconds?: number | null;
  question_results: QuestionResult[];
  mastery_updates?: MasteryNodeUpdate[];
  strengths_recorded?: string[];
  weaknesses_recorded?: string[];
  unlocked_nodes?: string[];
  overall_feedback: string;
  mastery_delta: number;
}

export interface QuizAttemptSummary {
  id: string;
  profile_id: string;
  roadmap_node_id?: string | null;
  /** Resolved server-side from the roadmap node, for labelling history. */
  topic_title?: string | null;
  mode: string;
  started_at: string;
  completed_at?: string | null;
  score?: number | null;
  total_questions: number;
  correct_count?: number | null;
}

export const quizApi = {
  generateQuiz: async (
    profileId: string,
    payload: {
      roadmap_node_id?: string;
      topic_title?: string;
      question_types?: QuestionType[];
      mode?: QuizMode;
      question_count?: number;
      time_limit_seconds?: number;
      difficulty?: DifficultyLevel;
    }
  ): Promise<QuizResponse> => {
    const res = await apiClient.post<ApiEnvelope<QuizResponse>>(
      `/profiles/${profileId}/quiz/generate`,
      payload
    );
    return res.data.data;
  },

  generateAssessment: async (
    profileId: string,
    payload: {
      assessment_type?: AssessmentType;
      roadmap_node_ids?: string[];
      chapter_title?: string;
      topic_titles?: string[];
      difficulty?: DifficultyLevel;
      question_count?: number;
      time_limit_seconds?: number;
    }
  ): Promise<QuizResponse> => {
    const res = await apiClient.post<ApiEnvelope<QuizResponse>>(
      `/profiles/${profileId}/quiz/assessments`,
      payload
    );
    return res.data.data;
  },

  submitQuiz: async (
    profileId: string,
    attemptId: string,
    payload: { answers: QuizUserAnswer[] }
  ): Promise<QuizResultResponse> => {
    const res = await apiClient.post<ApiEnvelope<QuizResultResponse>>(
      `/profiles/${profileId}/quiz/${attemptId}/submit`,
      payload
    );
    return res.data.data;
  },

  getHistory: async (profileId: string): Promise<QuizAttemptSummary[]> => {
    const res = await apiClient.get<ApiEnvelope<QuizAttemptSummary[]>>(
      `/profiles/${profileId}/quiz/history`
    );
    return res.data.data;
  },

  getResults: async (profileId: string, attemptId: string): Promise<QuizResultResponse> => {
    const res = await apiClient.get<ApiEnvelope<QuizResultResponse>>(
      `/profiles/${profileId}/quiz/${attemptId}/results`
    );
    return res.data.data;
  },
};

File: backend/app/schemas/quiz.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

QuestionType = Literal["mcq", "numerical", "short_answer"]
QuizMode = Literal["practice", "timed_assessment"]
AssessmentType = Literal["topic_assessment", "chapter_test", "comprehensive_exam", "diagnostic_test"]
DifficultyLevel = Literal["easy", "medium", "hard", "adaptive"]
ErrorClassification = Literal[
    "correct",
    "conceptual_misunderstanding",
    "calculation_error",
    "incomplete_answer",
    "misread_question",
]


class QuizOption(BaseModel):
    id: str = Field(..., description="Option key, e.g. A, B, C, D")
    text: str = Field(..., description="Option choice text")


class QuizQuestion(BaseModel):
    id: str
    question_type: QuestionType
    prompt: str
    options: list[QuizOption] | None = None
    target_value: float | None = None
    tolerance: float | None = 0.05
    rubric: str | None = None
    correct_answer: str | None = None
    explanation: str | None = None
    topic_title: str | None = None
    difficulty: DifficultyLevel = "medium"


class QuizQuestionPublic(BaseModel):
    id: str
    question_type: QuestionType
    prompt: str
    options: list[QuizOption] | None = None
    topic_title: str | None = None
    difficulty: DifficultyLevel = "medium"


class QuizGenerateRequest(BaseModel):
    roadmap_node_id: str | None = Field(None, description="Optional roadmap node ID to generate questions for")
    topic_title: str | None = Field(None, description="Optional topic or chapter title if node ID not provided")
    question_types: list[QuestionType] | None = Field(
        default_factory=lambda: ["mcq", "numerical", "short_answer"],
        description="Types of questions to generate",
    )
    mode: QuizMode = Field("practice", description="practice or timed_assessment")
    question_count: int = Field(5, ge=1, le=25, description="Number of questions to generate")
    time_limit_seconds: int | None = Field(None, ge=30, le=7200, description="Time limit for timed assessment mode")
    difficulty: DifficultyLevel = Field("medium", description="Difficulty level: easy, medium, hard, adaptive")


class AssessmentGenerateRequest(BaseModel):
    assessment_type: AssessmentType = Field("topic_assessment", description="Assessment format")
    roadmap_node_ids: list[str] | None = Field(None, description="List of roadmap node IDs covered")
    chapter_title: str | None = Field(None, description="Chapter title for chapter test")
    topic_titles: list[str] | None = Field(None, description="Custom list of topic titles")
    difficulty: DifficultyLevel = Field("medium", description="Difficulty level: easy, medium, hard, adaptive")
    question_count: int = Field(6, ge=1, le=30, description="Number of questions")
    time_limit_seconds: int | None = Field(None, ge=30, le=7200, description="Optional time limit")


class QuizUserAnswer(BaseModel):
    question_id: str
    user_answer: str


class QuizSubmitRequest(BaseModel):
    answers: list[QuizUserAnswer] = Field(default_factory=list)


class QuestionResult(BaseModel):
    question_id: str
    question_type: str
    prompt: str
    user_answer: str
    correct_answer: str
    is_correct: bool
    score: float = Field(..., ge=0.0, le=1.0)
    error_category: ErrorClassification = "correct"
    feedback: str = ""
    explanation: str = ""
    remediation_advice: str = ""


class MasteryNodeUpdate(BaseModel):
    node_id: str
    node_title: str
    previous_mastery: float
    new_mastery: float
    status: str
    unlocked: bool = True


class QuizResponse(BaseModel):
    id: str
    profile_id: str
    roadmap_node_id: str | None = None
    mode: str
    total_questions: int
    time_limit_seconds: int | None = None
    difficulty: str = "medium"
    questions: list[QuizQuestionPublic]


class QuizResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    roadmap_node_id: str | None = None
    mode: str
    started_at: datetime
    completed_at: datetime | None = None
    score: float = 0.0
    letter_grade: str = "N/A"
    total_questions: int
    correct_count: int = 0
    time_limit_seconds: int | None = None
    question_results: list[QuestionResult] = Field(default_factory=list)
    mastery_updates: list[MasteryNodeUpdate] = Field(default_factory=list)
    strengths_recorded: list[str] = Field(default_factory=list)
    weaknesses_recorded: list[str] = Field(default_factory=list)
    unlocked_nodes: list[str] = Field(default_factory=list)
    overall_feedback: str = ""
    mastery_delta: float = 0.0


class QuizAttemptSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    roadmap_node_id: str | None = None
    # Resolved from the roadmap node so history can be labelled by subject
    # rather than by a fragment of the attempt id.
    topic_title: str | None = None
    mode: str
    started_at: datetime
    completed_at: datetime | None = None
    score: float | None = None
    total_questions: int
    correct_count: int | None = None
